runner_test.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+// Use `go test -tree` to debug DumpRunner tests.
+var printTree = flag.Bool("tree", false, "print file tree in DumpRunner tests")
+
+func TestDumpRunner(t *testing.T) {
+	dir, err := ioutil.TempDir("", "test-dumprunner-dir-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	if *printTree {
+		defer func() { t.Log("\n" + tree(dir)) }()
+	}
+
+	// We will use dr several times to make sure it does what it should do.
+	// Each subtest begins with setting the cmd we will run.
+	dr := NewDumpRunner(dir)
+
+	// #1 -- common case, command exits with 0, output is saved to a file.
+	cmd := helperCommand("echo", "ok")
+
+	// Set cmd.Stdout to later ensure that dr.Run will write to both b and a
+	// file.
+	var b bytes.Buffer
+	cmd.Stdout = &b
+
+	// These are the files and contents we expect to see:
+	names := []string{"out"}
+	contents := []string{cmd.Args[len(cmd.Args)-1] + "\n"}
```

Comment:
Once I had this hard-coded, perhaps it was easier to read...

```go
	contents := []string{"ok\n"}
```

> would this be easier as: `contents := []string{cmd.Args[1:]}`?

That wouldn't compile, as it is attempting to put a slice where we expect a string.

And `cmd.Args[1:]` is not `[]string{"ok"}` as one may imagine, because `helperCommand` returns a `*exec.Cmd` that runs the test binary itself with crafted arguments, among those `"echo"` and `"ok"`.

We also need to include the newline character written to stdout.